File: backend/src/main/java/com/markethub/features/products/models/Product.java
package com.markethub.features.products.models;

import com.markethub.features.categories.models.Category;
import com.markethub.features.tags.models.Tag;
import com.markethub.features.user.models.User;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "products")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "seller_id", nullable = false)
    private User seller;

    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    private BigDecimal price;

    @Enumerated(EnumType.STRING)
    private ProductStatus status;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToMany
    @JoinTable(name = "product_tags", joinColumns = @JoinColumn(name = "product_id"), inverseJoinColumns = @JoinColumn(name = "tag_name"))
    private List<Tag> tags;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<ProductExportCountry> exportCountries;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: backend/internal/modules/products/models/product.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"

	"github.com/vendrya/markethub/internal/modules/user/models"
)

type Status string

const (
	StatusActive Status = "ACTIVE"
	StatusPaused Status = "PAUSED"
	StatusSold   Status = "SOLD"
)

type Product struct {
	ID          uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey"`
	Title       string    `gorm:"not null"`
	Description string    `gorm:"type:text;"`
	Price       float64   `gorm:"not null"`
	Status      Status    `gorm:"type:varchar(20);default:'ACTIVE'"`

	CategoryID uuid.UUID `json:"-" gorm:"type:uuid;not null"`
	Category   *Category `json:"category,omitempty"`

	UserID uuid.UUID `json:"-" gorm:"type:uuid;not null"`

	Tags []Tag       `gorm:"many2many:product_tags"`
	User models.User `json:"author,omitempty" gorm:"foreignKey:UserID"`

	CreatedAt time.Time
	UpdatedAt time.Time
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}
